checks/raw/shell_download_validate.go

```diff
@@ -696,6 +697,83 @@ func isChocoUnpinnedDownload(cmd []string) bool {
 	return true
 }
 
+func isUnpinnedNugetCliInstall(cmd []string) bool {
+	// Search for nuget commands.
+	if !isBinaryName("nuget", cmd[0]) {
+		return false
+	}
+
+	// Search for install commands.
+	if !strings.EqualFold(cmd[1], "install") {
+		return false
+	}
+
+	// package.config schema has required version field
+	// https://learn.microsoft.com/en-us/nuget/reference/packages-config#schema
+	if strings.HasSuffix(cmd[2], "packages.config") {
```

Comment:
does this file ensure that packages are declared immutably? In these examples https://learn.microsoft.com/en-us/nuget/reference/packages-config#schema, it seems like they are declared using a version. Are versions immutable in nuget?

Reply:
yes, checked with @JonDouglas a Nuget PM that Nuget follows semantic versioning 2.0.0, and versions cannot be changed. documented in the code. 
https://learn.microsoft.com/en-us/nuget/concepts/package-versioning#semantic-versioning-200

Reply:
so the NuGet registry enforces immutability on the content of a package? Say, a package maintainer publishes benign package `hello-v1.2.3`. Can they later push the package / version with a different hash (this time containing malicious content)?

Several ecosystems have a concept of "lock file", which "pins" packages to the cryptographic hash of the package content. In the case of NuGet, we trust the registry to enforce immutability?

Reply:
The central registry (NuGet.org) enforces immutability on package metadata such as the name, version, dependencies, etc. Once a package is published, its metadata cannot be changed. The package contents are generally seen as immutable once packaged and published. But like other ecosystems, it is not impossible for the contents to be overwritten if following bad practices on a private feed they control. [The central registry has a policy on deletion and unlisting for this purpose.](https://learn.microsoft.com/en-us/nuget/nuget-org/policies/deleting-packages)

In the case you outline, the maintainer would need to increment a new package version to then push with new contents. Usually through [semantic versioning](https://semver.org/).

We also support lock files, dependency pinning, and much more for cases where you may require more determinism.

Reply:
Got it, thanks for the information. Do you know if we should encourage lock files for applications? We do this for other ecosystems. I don't know what the best practices are for NuGet. But iiuc, recommending users to use lock files would provide similar guarantees while reducing trust on the registry. In terms of usability for users, is it the same amount of work?

Reply:
In OpenSSF we encourage folks to use lock files. But I want to be sure this is not going against NuGet's best practices.
Do you think we should do the same for NuGet, ie lock files? If it's as easy for users to use, we could recommend it instead of version pinning.

Wdut?

Reply:
is this meant as a documentation recommendation or in the scorecard checks? if in code, could you please link to where that is implemented fro other ecosystems? @laurentsimon

Reply:
sure, you're actually in the right file.
pip: https://github.com/ossf/scorecard/blob/main/checks/raw/shell_download_validate.go#L549. `--require-hashes` means there must be a lock file that is complete.
npm: https://github.com/ossf/scorecard/blob/main/checks/raw/shell_download_validate.go#L411-L415. `npm ci` command fails if there is no lock file or if it's incomplete.

We also check for similar immutable declaration for Dockerfile's BASE and GitHub Action `@<ref>` https://github.com/ossf/scorecard/blob/main/checks/raw/pinned_dependencies.go#L37-L42

Reply:
`RestorePackagesWithLockFile` alone does not seem enough, iiuc. `If NuGet detects a change in the defined dependencies as enumerated in the project file(s), it re-evaluates the package graph and updates the lock file`

Looks like there are various ways then
1. check the relevant file for `RestoreLockedMode` property being set. This may be dynamically set, so we'd need to also check for the propoerty it refers to MSBuild property set (`ContinuousIntegrationBuild ` in the article). On GitHub, there is an env variable `CI=true` which is set, so I suppose we'd verify the property is set with it?
2. for donet, check for `restore --locked-mode`. If another command is used (`install` or `restore` alone), we would flag it
3. for msbuild, same approach as (2) but with `/t:restore /p:RestoreLockedMode=true`
4. I'm not sure about nugget. Do you know if it has similar support?

Since there are many possibilities, let's split the PR to make it easier to merge one-by-one. 

Maybe we could start with (2) and (3) if you agree they are the right approaches to recommend?

Wdut?

Reply:
> I'm all in favor for breaking up the issue to a number of prs, thanks for suggesting that! I can open the open points as g.h issues after we lock down the plan.

+1, thanks!

> Should there be a lower score that is not zero for using pinned versions without lock file (I.e. As the current implementation) and the highest score for repos that do use lock file? 

Definitely worth considering. If there's an easy path for upgrade for users, I think it would be OK. We just want to avoid a situation where we ask users to do a bunch of work for version, and the work is not "re-usable" to further improve for hashes.

Reply:
msbuild & dotnet restore are exactly the same, just different commands. 
NuGet.exe also has matching commands for lock files. 

The expectation is that users run `--locked-mode` on their CI when using lock files.

Note that lock files are not really possible for what seems to be referred as pinned dependencies, as for example with nuget.exe install, there's simply no project file in play.

Reply:
Another note is that NuGet's resolution (in PackageReference) explicitly prefers lowest acceptable version, so it's not all that likely that restore today is different from restore 1 month from now even without a lock file, given the policy of nuget.org.

Reply:
LGTM. So we will keep this PR and tweak the score not be lower than the full score?

Reply:
yes, i guess that will be more relevant once #2690 is merged and more dependency types can be scored differently

Reply:
@gabibguti please take this thread into consideration in your PR https://github.com/ossf/scorecard/pull/2690

Reply:
Thanks everyone. Resolving comment.